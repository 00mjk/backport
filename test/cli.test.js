const inquirer = require('inquirer');
const axios = require('axios');
const nock = require('nock');
const httpAdapter = require('axios/lib/adapters/http');
const { init } = require('../src/cli');
const github = require('../src/github');
const commitsMock = require('./mocks/commits.json');
const { mockBackportDirPath } = require('./testHelpers');
const utils = require('../src/utils');

const host = 'http://localhost';

axios.defaults.host = host;
axios.defaults.adapter = httpAdapter;

describe('select commit that originated from pull request', () => {
  const owner = 'elastic';
  const repoName = 'backport-cli-test';
  const fullRepoName = `${owner}/${repoName}`;

  beforeEach(() => {
    mockBackportDirPath();
    jest.spyOn(utils, 'exec').mockReturnValue(Promise.resolve());

    jest
      .spyOn(inquirer, 'prompt')
      .mockReturnValueOnce(Promise.resolve({ fullRepoName }))
      .mockReturnValueOnce(
        Promise.resolve({
          commit: {
            message: 'myCommitMessage',
            sha: 'mySha'
          }
        })
      )
      .mockReturnValueOnce(
        Promise.resolve({
          version: 'myVersion'
        })
      );

    nock('https://api.github.com')
      .get(`/repos/${owner}/${repoName}/commits`)
      .query({ author: 'sqren', per_page: '5', access_token: 'myAccessToken' })
      .reply(200, commitsMock);

    nock('https://api.github.com')
      .get(`/search/issues`)
      .query({
        q: 'repo:elastic/backport-cli-test mySha',
        access_token: 'myAccessToken'
      })
      .reply(200, {
        items: [
          {
            number: 'myPullRequest'
          }
        ]
      });

    nock('https://api.github.com')
      .post(`/repos/${owner}/${repoName}/pulls`)
      .query({ access_token: 'myAccessToken' })
      .reply(200, {
        html_url: 'myHtmlUrl'
      });

    jest.spyOn(github, 'getCommits');
    jest.spyOn(github, 'createPullRequest');

    return init(
      {
        username: 'sqren',
        accessToken: 'myAccessToken',
        repositories: [
          {
            name: fullRepoName,
            versions: ['6.x', '6.0', '5.6', '5.5', '5.4']
          }
        ]
      },
      { cwd: '/my/path' }
    );
  });

  it('prompt should display list of repository names', () => {
    expect(inquirer.prompt).toHaveBeenCalledWith([
      {
        choices: ['elastic/backport-cli-test'],
        message: 'Select repository',
        name: 'fullRepoName',
        type: 'list'
      }
    ]);
  });

  it('getCommit should be called with correct args', () => {
    expect(github.getCommits).toHaveBeenCalledWith(
      'elastic',
      'backport-cli-test',
      'sqren'
    );
  });

  it('createPullRequest should be called with correct args', () => {
    expect(github.createPullRequest).toHaveBeenCalledWith(
      'elastic',
      'backport-cli-test',
      {
        base: 'myVersion',
        body: 'Backports pull request #myPullRequest to myVersion',
        head: 'sqren:backport/myVersion/pr-myPullRequest',
        title: '[Backport] myCommitMessage'
      }
    );
  });

  it('prompt calls should match snapshot', () => {
    expect(inquirer.prompt.mock.calls).toMatchSnapshot();
  });

  it('exec should be called with correct args', () => {
    expect(utils.exec.mock.calls).toMatchSnapshot();
  });
});

describe('select commit that originated from commit', () => {
  const owner = 'elastic';
  const repoName = 'backport-cli-test';
  const fullRepoName = `${owner}/${repoName}`;

  beforeEach(() => {
    mockBackportDirPath();
    jest.spyOn(utils, 'exec').mockReturnValue(Promise.resolve());

    jest
      .spyOn(inquirer, 'prompt')
      .mockReturnValueOnce(Promise.resolve({ fullRepoName }))
      .mockReturnValueOnce(
        Promise.resolve({
          commit: {
            message: 'myCommitMessage',
            sha: 'mySha'
          }
        })
      )
      .mockReturnValueOnce(
        Promise.resolve({
          version: 'myVersion'
        })
      );

    nock('https://api.github.com')
      .get(`/repos/${owner}/${repoName}/commits`)
      .query({ author: 'sqren', per_page: '5', access_token: 'myAccessToken' })
      .reply(200, commitsMock);

    nock('https://api.github.com')
      .get(`/search/issues`)
      .query({
        q: 'repo:elastic/backport-cli-test mySha',
        access_token: 'myAccessToken'
      })
      .reply(200, {
        items: []
      });

    nock('https://api.github.com')
      .post(`/repos/${owner}/${repoName}/pulls`)
      .query({ access_token: 'myAccessToken' })
      .reply(200, {
        html_url: 'myHtmlUrl'
      });

    jest.spyOn(github, 'getCommits');
    jest.spyOn(github, 'createPullRequest');

    return init(
      {
        username: 'sqren',
        accessToken: 'myAccessToken',
        repositories: [
          {
            name: fullRepoName,
            versions: ['6.x', '6.0', '5.6', '5.5', '5.4']
          }
        ]
      },
      { cwd: '/my/path' }
    );
  });

  it('createPullRequest should be called with correct args', () => {
    expect(github.createPullRequest).toHaveBeenCalledWith(
      'elastic',
      'backport-cli-test',
      {
        base: 'myVersion',
        body: 'Backports commit mySha to myVersion',
        head: 'sqren:backport/myVersion/commit-mySha',
        title: '[Backport] myCommitMessage'
      }
    );
  });
});
