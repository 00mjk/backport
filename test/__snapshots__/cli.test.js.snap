// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli prompt calls should match snapshot 1`] = `
Array [
  Array [
    Array [
      Object {
        "choices": Array [
          "elastic/backport-cli-test",
        ],
        "message": "Select repository",
        "name": "fullRepoName",
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "Update readme: Make ui-framework url a link",
            "short": "Update readme: Make ui-framework url a link",
            "value": Object {
              "date": "2017-09-05T08:48:42Z",
              "message": "Update readme: Make ui-framework url a link",
              "reference": Object {
                "type": "commit",
                "value": "7939a98",
              },
              "sha": "7939a98aadb94ead30b54bea7974cd02dc816b02",
            },
          },
          Object {
            "name": "Update react 15.6.1 (#13672)",
            "short": "Update react 15.6.1 (#13672)",
            "value": Object {
              "date": "2017-08-28T08:27:51Z",
              "message": "Update react 15.6.1 (#13672)",
              "reference": Object {
                "type": "pullRequest",
                "value": "13672",
              },
              "sha": "52552a93efa5b4038ea57d9cb3d158ba6a1ff71d",
            },
          },
          Object {
            "name": "Mock out static files when running in Jest (#13315)",
            "short": "Mock out static files when running in Jest (#13315)",
            "value": Object {
              "date": "2017-08-07T15:35:59Z",
              "message": "Mock out static files when running in Jest (#13315)",
              "reference": Object {
                "type": "pullRequest",
                "value": "13315",
              },
              "sha": "8cfbb82cfb50aa962410a9682d96335a252d751a",
            },
          },
        ],
        "message": "Select a commit to backport",
        "name": "commit",
        "type": "list",
      },
      Object {
        "choices": Array [
          "6.x",
          "6.0",
          "5.6",
          "5.5",
          "5.4",
        ],
        "message": "Select version you want to backport to",
        "name": "version",
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "elastic/backport-cli-test",
        ],
        "message": "Select repository",
        "name": "fullRepoName",
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "Update readme: Make ui-framework url a link",
            "short": "Update readme: Make ui-framework url a link",
            "value": Object {
              "date": "2017-09-05T08:48:42Z",
              "message": "Update readme: Make ui-framework url a link",
              "reference": Object {
                "type": "commit",
                "value": "7939a98",
              },
              "sha": "7939a98aadb94ead30b54bea7974cd02dc816b02",
            },
          },
          Object {
            "name": "Update react 15.6.1 (#13672)",
            "short": "Update react 15.6.1 (#13672)",
            "value": Object {
              "date": "2017-08-28T08:27:51Z",
              "message": "Update react 15.6.1 (#13672)",
              "reference": Object {
                "type": "pullRequest",
                "value": "13672",
              },
              "sha": "52552a93efa5b4038ea57d9cb3d158ba6a1ff71d",
            },
          },
          Object {
            "name": "Mock out static files when running in Jest (#13315)",
            "short": "Mock out static files when running in Jest (#13315)",
            "value": Object {
              "date": "2017-08-07T15:35:59Z",
              "message": "Mock out static files when running in Jest (#13315)",
              "reference": Object {
                "type": "pullRequest",
                "value": "13315",
              },
              "sha": "8cfbb82cfb50aa962410a9682d96335a252d751a",
            },
          },
        ],
        "message": "Select a commit to backport",
        "name": "commit",
        "type": "list",
      },
      Object {
        "choices": Array [
          "6.x",
          "6.0",
          "5.6",
          "5.5",
          "5.4",
        ],
        "message": "Select version you want to backport to",
        "name": "version",
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "elastic/backport-cli-test",
        ],
        "message": "Select repository",
        "name": "fullRepoName",
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "Update readme: Make ui-framework url a link",
            "short": "Update readme: Make ui-framework url a link",
            "value": Object {
              "date": "2017-09-05T08:48:42Z",
              "message": "Update readme: Make ui-framework url a link",
              "reference": Object {
                "type": "commit",
                "value": "7939a98",
              },
              "sha": "7939a98aadb94ead30b54bea7974cd02dc816b02",
            },
          },
          Object {
            "name": "Update react 15.6.1 (#13672)",
            "short": "Update react 15.6.1 (#13672)",
            "value": Object {
              "date": "2017-08-28T08:27:51Z",
              "message": "Update react 15.6.1 (#13672)",
              "reference": Object {
                "type": "pullRequest",
                "value": "13672",
              },
              "sha": "52552a93efa5b4038ea57d9cb3d158ba6a1ff71d",
            },
          },
          Object {
            "name": "Mock out static files when running in Jest (#13315)",
            "short": "Mock out static files when running in Jest (#13315)",
            "value": Object {
              "date": "2017-08-07T15:35:59Z",
              "message": "Mock out static files when running in Jest (#13315)",
              "reference": Object {
                "type": "pullRequest",
                "value": "13315",
              },
              "sha": "8cfbb82cfb50aa962410a9682d96335a252d751a",
            },
          },
        ],
        "message": "Select a commit to backport",
        "name": "commit",
        "type": "list",
      },
      Object {
        "choices": Array [
          "6.x",
          "6.0",
          "5.6",
          "5.5",
          "5.4",
        ],
        "message": "Select version you want to backport to",
        "name": "version",
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "elastic/backport-cli-test",
        ],
        "message": "Select repository",
        "name": "fullRepoName",
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "Update readme: Make ui-framework url a link",
            "short": "Update readme: Make ui-framework url a link",
            "value": Object {
              "date": "2017-09-05T08:48:42Z",
              "message": "Update readme: Make ui-framework url a link",
              "reference": Object {
                "type": "commit",
                "value": "7939a98",
              },
              "sha": "7939a98aadb94ead30b54bea7974cd02dc816b02",
            },
          },
          Object {
            "name": "Update react 15.6.1 (#13672)",
            "short": "Update react 15.6.1 (#13672)",
            "value": Object {
              "date": "2017-08-28T08:27:51Z",
              "message": "Update react 15.6.1 (#13672)",
              "reference": Object {
                "type": "pullRequest",
                "value": "13672",
              },
              "sha": "52552a93efa5b4038ea57d9cb3d158ba6a1ff71d",
            },
          },
          Object {
            "name": "Mock out static files when running in Jest (#13315)",
            "short": "Mock out static files when running in Jest (#13315)",
            "value": Object {
              "date": "2017-08-07T15:35:59Z",
              "message": "Mock out static files when running in Jest (#13315)",
              "reference": Object {
                "type": "pullRequest",
                "value": "13315",
              },
              "sha": "8cfbb82cfb50aa962410a9682d96335a252d751a",
            },
          },
        ],
        "message": "Select a commit to backport",
        "name": "commit",
        "type": "list",
      },
      Object {
        "choices": Array [
          "6.x",
          "6.0",
          "5.6",
          "5.5",
          "5.4",
        ],
        "message": "Select version you want to backport to",
        "name": "version",
        "type": "list",
      },
    ],
  ],
]
`;

exports[`cli utils,  1`] = `
Array [
  Array [
    "git reset --hard && git checkout master && git pull origin master",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git branch backport/myVersion/pr-myPullRequest origin/myVersion --force && git checkout backport/myVersion/pr-myPullRequest ",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git push sqren backport/myVersion/pr-myPullRequest --force",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git reset --hard && git checkout master && git pull origin master",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git branch backport/myVersion/pr-myPullRequest origin/myVersion --force && git checkout backport/myVersion/pr-myPullRequest ",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git push sqren backport/myVersion/pr-myPullRequest --force",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git reset --hard && git checkout master && git pull origin master",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git branch backport/myVersion/pr-myPullRequest origin/myVersion --force && git checkout backport/myVersion/pr-myPullRequest ",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git push sqren backport/myVersion/pr-myPullRequest --force",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git reset --hard && git checkout master && git pull origin master",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git branch backport/myVersion/pr-myPullRequest origin/myVersion --force && git checkout backport/myVersion/pr-myPullRequest ",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git push sqren backport/myVersion/pr-myPullRequest --force",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git reset --hard && git checkout master && git pull origin master",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git branch backport/myVersion/pr-myPullRequest origin/myVersion --force && git checkout backport/myVersion/pr-myPullRequest ",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git push sqren backport/myVersion/pr-myPullRequest --force",
    Object {
      "cwd": "/srv/backport-cli/test/.backport/repositories/elastic/backport-cli-test",
    },
  ],
]
`;
